sdk/include/opentelemetry/sdk/metrics/observer_result.h

```diff
@@ -0,0 +1,44 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+#ifndef ENABLE_METRICS_PREVIEW
+#  include "opentelemetry/common/key_value_iterable.h"
+#  include "opentelemetry/metrics/observer_result.h"
+#  include "opentelemetry/sdk/metrics/state/attributes_hashmap.h"
+#  include "opentelemetry/sdk/metrics/view/attributes_processor.h"
+
+#  include <map>
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+namespace metrics
+{
+template <class T>
+class ObserverResult final : public opentelemetry::metrics::ObserverResult<T>
+{
+public:
+  ObserverResult(const AttributesProcessor *attributes_processor)
+      : attributes_processor_(attributes_processor)
+  {}
+
+  virtual void Observe(T value) noexcept override { data_.insert({{}, value}); }
+
+  virtual void Observer(T value,
+                        const opentelemetry::common::KeyValueIterable &attributes) noexcept override
+  {
+    auto attr = attributes_processor_->process(attributes);
+    data_.insert({attr, value});
+  }
+
+  std::unordered_map<MetricAttributes, T, AttributeHashGenerator> data_;
```

Comment:
No, ideally this shouldn't be public. As this class is tightly coupled with AsyncMetricStorage, I kept it public to avoid an extra function call during collection. Have added a getter method for data now, and made it private. Thanks for noticing it.